File: src/main/java/com/garageos/modules/estimateitem/service/EstimateItemService.java
package com.garageos.modules.estimateitem.service;

import com.garageos.modules.estimateitem.dto.request.CreateEstimateItemRequest;
import com.garageos.modules.estimateitem.dto.response.EstimateItemResponse;

import java.util.List;

public interface EstimateItemService {

    EstimateItemResponse addItem(
            Long estimateId,
            CreateEstimateItemRequest request);

    EstimateItemResponse getItem(Long id);

    List<EstimateItemResponse> getItems(Long estimateId);

    List<EstimateItemResponse> getItemsByComplaint(Long complaintId);

    EstimateItemResponse updateItem(
            Long id,
            CreateEstimateItemRequest request);

    void deleteItem(Long id);

    /**
     * Mission: customer select/deselect of individual estimate items,
     * before approval. Customer-only, for an item on their own estimate,
     * only while the estimate has not yet been approved - see
     * EstimateItemServiceImpl.setItemSelection for the exact checks.
     */
    EstimateItemResponse setItemSelection(Long itemId, boolean selected);

}

File: src/main/resources/static/services/estimateItemService.js
window.EstimateItemService = {

    async addItem(estimateId, request) {

        return await Api.post(

            `/estimates/${estimateId}/items`,

            request

        );

    },

    async getItems(estimateId) {

        return await Api.get(

            `/estimates/${estimateId}/items`

        );

    },

    async updateItem(id, request) {

        return await Api.put(

            `/estimate-items/${id}`,

            request

        );

    },

    async deleteItem(id) {

        return Api.delete(

            `/estimate-items/${id}`

        );

    }

};